nextstrain/cli/util.py

```diff
@@ -213,3 +213,56 @@ def resolve_path(path: Path) -> Path:
         return path.resolve(strict = True) # type: ignore
     else:
         return path.resolve()
+
+
+def byte_quantity(quantity: str) -> int:
+    """
+    Parses a string *quantity* consisting of a number, optional whitespace, and
+    a unit of bytes.
+
+    Returns the number of bytes in *quantity*, as an integer.
+
+    Supported units:
+
+    * ``b`` (bytes)
+    * ``kb`` (kilobytes)
+    * ``mb`` (megabytes)
+    * ``gb`` (gigabytes)
+    * ``kib`` (kibibytes)
+    * ``mib`` (mebibytes)
+    * ``gib`` (gibibytes)
+
+    Units are not case sensitive.  If no unit is given, bytes is assumed.
+
+    Raises a :py:class:`ValueError` if *quantity* is not parseable.
+
+    >>> byte_quantity("2Kb")
+    2000
+    >>> byte_quantity("2 kib")
+    2048
+    >>> byte_quantity("1.5GB")
+    1500000000
+    >>> byte_quantity("1024")
+    1024
+    """
+    match = re.search(r'(?xi)^ (\S+?) \s* ((?:(?:[kmg]i?)?b)?) $', quantity.strip())
```

Comment:
What is going on here :laughing: 
I understand the second half, `((?:(?:...` : the code is doing a look ahead for k/m/g (i)(b) suffixes without matching on them. But I am struggling to understand the function of `(?xi)^ (\S+?)`.

Reply:
`(?:…)` is a non-capturing group, not a look-ahead. Those non-capturing groups are surrounded by a capturing group `(…)`.

`(?xi)` sets the flags for the regex. It must be the first thing ("regex atom") when its used. `x` is the `re.X` or `re.VERBOSE` flag. `i` is the `re.I` or `re.IGNORECASE` flag. I'm used the short flags from decades of using them, but perhaps I should expand them for clarity here and move them into arguments to `re.search` instead of part of the pattern. (I like them as part of the pattern because they tell me upfront how to read the pattern, instead of coming afterwards, but this is a minor thing overall.)

`(\S+?)` is not an optional match, but a shortest ("ungreedy") match of one or more non-whitespace chars. `?` modifies a quantifier, in this case `+`, to change its behaviour from longest ("greedy") match to shortest ("ungreedy") match. This ensures that the `\S+` atom doesn't eat up the unit specifier in a string without spaces, like `18gb`.

Would this be better?

```python
re.search(
    r"""
    ^                           # start of string
    (?P<value> \S+? )           # the numeric value: shortest match of non-whitespace chars
    \s*                         # optional whitespace between the value and the unit
    (?P<unit>
     (?:
      (?: [kmg]i? )? b)? )      # the unit: [kmg]b, [kmg]ib, b, or nothing
    $                           # end of the string
    """,
    quantity.strip(),
    re.VERBOSE | re.IGNORECASE)
```

or maybe

```python
re.search(
    r"""
    ^                           # start of string
    (?P<value> \S+? )           # the numeric value: shortest match of non-whitespace chars
    \s*                         # optional whitespace between the value and the unit
    (?P<unit>                   # the unit:
          [kmg]b                #    kb, mb, gb
        | [kmg]ib               #    kib, mib, gib
        | b                     #    b
    )?                          # or none of the above, in which case we assume bytes.
    $                           # end of the string
    """,
    quantity.strip(),
    re.VERBOSE | re.IGNORECASE)
```

Reply:
Testing on https://regex101.com it currently matches strings like `18kmb` with `18k` being the `value`, but I guess `float(value)` would raise a ValueError in this case?

Reply:
I thought I understood regex fairly well until this PR. I did not even realize python had regex flags you could set at the beginning of the expression.

Thank you for the explanations. Jover also pointed me to regex101.com which I will use in the future for understanding complex expressions. 

As for documenting the pieces of the regex expression you wrote, I think the usefulness is pretty subjective. For example, `^` and `$` are very straightforward for me and don't need an explanation, but others may disagree. I personally think explaining the more complex pieces of the regex is useful. Or I could just turn to regex101.com in the future.

Reply:
Yeah, it relies on the float conversion to raise errors with the value part. This seemed easier than representing everything understood by `float()` as a regex, e.g. `12e3 mb` or `.5gb`. Maybe it should be stricter though.

Reply:
I repushed this with something I think is at least a little more clear.